src/libraries/Common/src/System/Net/Http/Http3/Helpers/VariableLengthIntegerHelper.cs

```diff
@@ -0,0 +1,214 @@
+
+                switch (firstByte & LengthMask)
+                {
+                    case LengthOneByte:
+                        value = firstByte;
+                        bytesRead = 1;
+                        return true;
+                    case LengthTwoByte:
+                        if (BinaryPrimitives.TryReadUInt16BigEndian(buffer, out ushort serializedShort))
+                        {
+                            value = serializedShort - TwoByteSubtract;
+                            bytesRead = 2;
+                            return true;
+                        }
+                        break;
+                    case LengthFourByte:
+                        if (BinaryPrimitives.TryReadUInt32BigEndian(buffer, out uint serializedInt))
+                        {
+                            value = serializedInt - FourByteSubtract;
+                            bytesRead = 4;
+                            return true;
+                        }
+                        break;
+                    default: // LengthEightByte
+                        Debug.Assert((firstByte & LengthMask) == LengthEightByte);
+                        if (BinaryPrimitives.TryReadUInt64BigEndian(buffer, out ulong serializedLong))
+                        {
+                            value = (long)(serializedLong - EightByteSubtract);
+                            Debug.Assert(value >= 0 && value <= EightByteLimit, "Serialized values are within [0, 2^62).");
+
+                            bytesRead = 8;
+                            return true;
+                        }
+                        break;
+                }
+            }
+
+            value = 0;
+            bytesRead = 0;
+            return false;
+        }
+
+        public static bool TryRead(ref SequenceReader<byte> reader, out long value)
+        {
+            // Hot path: we probably have the entire integer in one unbroken span.
+            if (TryRead(reader.UnreadSpan, out value, out int bytesRead))
+            {
+                reader.Advance(bytesRead);
+                return true;
+            }
+
+            // Cold path: copy to a temporary buffer before calling span-based read.
+            return TryReadSlow(ref reader, out value);
+
+            static bool TryReadSlow(ref SequenceReader<byte> reader, out long value)
+            {
+                ReadOnlySpan<byte> span = reader.CurrentSpan;
+
+                if (reader.TryPeek(out byte firstByte))
+                {
+                    int length =
+                        (firstByte & LengthMask) switch
+                        {
+                            LengthOneByte => 1,
+                            LengthTwoByte => 2,
+                            LengthFourByte => 4,
+                            _ => 8 // LengthEightByte
+                    };
+
+                    Span<byte> temp = (stackalloc byte[8])[..length];
+                    if (reader.TryCopyTo(temp))
```

Comment:
If the `SequenceReader<T>` contains only partial data, is it possible that `TryCopyTo` will succeed but won't fill the entire buffer, then the recursive call to `TryRead` below will start reading potentially uninitialized stackalloced memory? Could be a subtle information disclosure leak.

Reply:
Maybe my understanding of `SequenceReader.TryCopyTo` is wrong -- won't `TryCopyTo` fail if it can't copy the entire length?

(also -- this isn't a recursive call, it's calling the `Span` overload)

Reply:
Example: The length marker says that there should be 2 bytes of data, but the SequenceReader contains only 1 byte of data. This routine stack-allocates a 2-byte buffer (contents uninitialized) then calls `SequenceReader.TryCopyTo`. The `TryCopyTo` method happily copies over the 1 byte of data it has, so now the original stack-allocated buffer contains a mix of initialized + uninitialized data. The partially-uninitialized buffer is then passed to a parsing routine. The behavior of the parsing routine may now be dependent on whatever garbage data was previously on the stack.

Reply:
Our documentation states of `TryCopyTo`'s return:
> true if there is enough data to completely fill the destination span; otherwise, false.

A glance at the implementation seems like this is correct.

Given this, the span will not have any uninitialized data in the event that it returns `true`.

Reply:
I relent - you're correct. It's just different than the other `CopyTo` methods in the framework and confused me.
Carry on, good sir. :)